File: src/main/java/svnserver/repository/locks/LockDescSerializer.java
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.repository.locks;

import org.jetbrains.annotations.NotNull;
import org.mapdb.DataInput2;
import org.mapdb.DataOutput2;
import org.mapdb.serializer.GroupSerializerObjectArray;

import java.io.IOException;

public final class LockDescSerializer extends GroupSerializerObjectArray<LockDesc> {

  @NotNull
  public static final LockDescSerializer instance = new LockDescSerializer();

  @Override
  public void serialize(@NotNull DataOutput2 out, @NotNull LockDesc value) throws IOException {
    out.writeUTF(value.getPath());

    if (value.getBranch() != null) {
      out.writeBoolean(true);
      out.writeUTF(value.getBranch());
    } else {
      out.writeBoolean(false);
    }

    if (value.getHash() != null) {
      out.writeBoolean(true);
      out.writeUTF(value.getHash());
    } else {
      out.writeBoolean(false);
    }

    out.writeUTF(value.getToken());
    out.writeUTF(value.getOwner());

    if (value.getComment() != null) {
      out.writeBoolean(true);
      out.writeUTF(value.getComment());
    } else {
      out.writeBoolean(false);
    }

    out.writeLong(value.getCreated());
  }

  @Override
  public LockDesc deserialize(@NotNull DataInput2 input, int available) throws IOException {
    final String path = input.readUTF();
    final String branch = input.readBoolean() ? input.readUTF() : null;
    final String hash = input.readBoolean() ? input.readUTF() : null;
    final String token = input.readUTF();
    final String owner = input.readUTF();
    final String comment = input.readBoolean() ? input.readUTF() : null;
    final long created = input.readLong();
    return new LockDesc(path, branch, hash, token, owner, comment, created);
  }

  @Override
  public int fixedSize() {
    return -1;
  }
}

File: src/main/kotlin/svnserver/repository/locks/LockDescSerializer.kt
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.repository.locks

import org.mapdb.DataInput2
import org.mapdb.DataOutput2
import org.mapdb.serializer.GroupSerializerObjectArray
import java.io.IOException

fun DataOutput2.writeNullableUTF(value: String?) {
    if (value == null) {
        writeBoolean(false)
    } else {
        writeBoolean(true)
        writeUTF(value)
    }
}

fun DataInput2.readNullableUTF() : String? {
    return if (readBoolean()) readUTF() else null
}

class LockDescSerializer : GroupSerializerObjectArray<LockDesc>() {
    @Throws(IOException::class)
    override fun serialize(out: DataOutput2, value: LockDesc) {
        out.writeUTF(value.path)
        out.writeNullableUTF(value.branch)
        out.writeNullableUTF(value.hash)
        out.writeUTF(value.token)
        out.writeNullableUTF(value.owner)
        out.writeNullableUTF(value.comment)
        out.writeLong(value.created)
    }

    @Throws(IOException::class)
    override fun deserialize(input: DataInput2, available: Int): LockDesc {
        val path: String = input.readUTF()
        val branch: String? = input.readNullableUTF()
        val hash: String? = input.readNullableUTF()
        val token: String = input.readUTF()
        val owner: String? = input.readNullableUTF()
        val comment: String? = input.readNullableUTF()
        val created: Long = input.readLong()
        return LockDesc(path, branch, hash, token, owner, comment, created)
    }

    override fun fixedSize(): Int {
        return -1
    }

    companion object {
        val instance: LockDescSerializer = LockDescSerializer()
    }
}
